### DISCUSSION TITLE
Issues with workspaces trying to deploy Fief with Postgres in GCP Cloud run

### DISCUSSION THREAD

**Discussion by dhivah** [fief-dev/fief#177] [Q&A]
I am trying to deploy Fief using Postgres DB in Google Cloud run. I have configured the Postgres database as recommended in the documentation. I have also configured the `FIEF_MAIN_USER_EMAIL` and `FIEF_MAIN_USER_PASSWORD`. I am able to load the site and login, but when I navigate to `/admin/` page it tries to create a new workspace and this new workspace is redirecting to a subdomain which is not what I wanted as sub domain redirection is not possible with Cloud Run. Is it possible to just use the default workspace, similar to the quickstart docker image that uses sqlite?

**Comment by frankie567** [comment]
Hi @dhivah 👋 

It looks like your main user is not tied to the main workspace. I would suggest to wipe the database and relaunch the Fief server. When starting, you should see something like this in the logs:

```
Main Fief workspace created
Main Fief user created
```

**Reply by dhivah** [reply]
@frankie567 I have created a new postgres database and tried running the cloud run but I am seeing the exact same issue. 
```
UniqueViolation: duplicate key value violates unique constraint
 "pg_type_typname_nsp_index"
DETAIL: Key (typname, typnamespace)=(alembic_version, 2200) already exists.
```
I see the `Main Fief workspace created` and then there is an error and then immediately I see `Main Fief workspace already exists` and `Main Fief user already exists`

```
 │ /usr/local/lib/python3.11/site-packages/fief/cli.py:381 in run_server │
 │ │
378 │ │ │ ) │
 379 │ │ │ │
 380 │ │ │ try: │
 ❱ 381 │ │ │ │ loop.run_until_complete( │
 382 │ │ │ │ │ create_main_fief_user(user_email, user_password) │
 383 │ │ │ │ ) │
 384 │ │ │ │ typer.echo("Main Fief user created") │
```
It looks like there is an issue with user creation. Is Redis connection required for the workspace and user creation?

**Reply by dhivah** [reply]
@frankie567 Here is the complete log file
[error_log.txt](https://github.com/fief-dev/fief/files/11361268/error_log.txt)

Thanks for your help

**Reply by frankie567** [reply]
Yes, the Redis server is needed for the server to work. By default, the Docker image starts both the Fief Server, worker and a Redis: https://github.com/fief-dev/fief/blob/9acbcd71f82c297f3ae33b52b5ad7c73a111a9f8/docker/supervisord.conf

If you override the run command, you'll have to come with your Redis server and set its URL via environment variable: https://docs.fief.dev/self-hosting/environment-variables/#redis

**Reply by dhivah** [reply]
@frankie567 could you please take a moment to check the error log file. There seems to be an issue with the database call too which i believe is not linked to redis. I created a new database and yet the issue persists.

**Reply by frankie567** [reply]
Yes, I did. So the SQL error is the following:

```
DEFAULT 2023-04-30T11:20:51.651798Z UniqueViolation: duplicate key value violates unique constraint
DEFAULT 2023-04-30T11:20:51.651802Z "pg_type_typname_nsp_index"
DEFAULT 2023-04-30T11:20:51.651814Z DETAIL: Key (typname, typnamespace)=(alembic_version, 2200) already exists.
```

It seems to be an error arising when [conflicting operations are done in parallel](https://stackoverflow.com/a/55286267/3864300). It's probably because Cloud Run tries to run several instances. I don't think it's a big deal since we see afterwards the container is started again and the error doesn't happen again.

However, we clearly see the error with the Redis server:

```
DEFAULT 2023-04-30T11:20:56.785604Z │ │ socket_address = ('127.0.0.1', 6379) │ │
DEFAULT 2023-04-30T11:20:56.785608Z │ │ socktype = <SocketKind.SOCK_STREAM: 1> │ │
DEFAULT 2023-04-30T11:20:56.785614Z │ ╰──────────────────────────────────────────────────────────────────────────╯ │
DEFAULT 2023-04-30T11:20:56.785620Z ╰──────────────────────────────────────────────────────────────────────────────╯
DEFAULT 2023-04-30T11:20:56.785629Z ConnectionRefusedError: [Errno 111] Connection refused
DEFAULT 2023-04-30T11:20:56.785637Z During handling of the above exception, another exception occurred:
DEFAULT 2023-04-30T11:20:56.785643Z ╭───────────────────── Traceback (most recent call last) ────────────
```

**Reply by dhivah** [reply]
@frankie567 I will set up the redis and let you know how it goes. thanks for looking into the logs.

**Reply by dhivah** [reply]
@frankie567 the main server runs after I configured the redis server. But I have issues with running the worker. GCP cloud run and cloud functions do not work as they do not handle background worker.  Is it recommended to run supervisord with 2 processes for server and worker in cloud run on Production? If not, can I configure the worker to be triggered every minute through a cronjob, would that work? other alternatives are to run cloud function that are triggered by some event, such as an trigger from cloud tasks. will that be possible to configure it with Fief?

**Reply by frankie567** [reply]
If you want to limit your costs, then it should be okay (even if not ideal) to run the server and the worker on the same Cloud Run instance using supervisord.

Ideally, you should spawn another Cloud Run instance to run the worker (so you have two running containers: one for the server, one for the worker).

Cloud Functions is not an option unfortunately, background tasks are implemented in Python with Dramatiq. Cloud Functions require to be defined with their own syntax/framework.

**Reply by dhivah** [reply]
I am trying to run it as a separate cloud run but there are limitations to cloud run. Cloud run caters more towards HTTP endpoints and Cloud run expects a container port which is used to check for health. I am unable to run the Cloud run with only the worker command. So any suggestions on how to run it?

**Reply by frankie567** [reply]
> I am trying to run it as a separate cloud run but there are limitations to cloud run. Cloud run caters more towards HTTP endpoints and Cloud run expects a container port which is used to check for health. I am unable to run the Cloud run with only the worker command. So any suggestions on how to run it?

Actually, the worker starts a simple HTTP server to expose Prometheus metrics on port 9191: https://dramatiq.io/advanced.html#prometheus-metrics

That should do the trick

**Reply by dhivah** [reply]
Thanks @frankie567 that worked and now the worker is running on a separate cloud run instance.

How do I add additional workspace admin user. I am able to login using the Main fief user configured in the CLI but I when I add additional user to the default tenant. The newly added account cannot access the admin portal and they are redirected to create new workspace page. I want to add few users who get admin access to the default workspace. How can this be done? I checked the admin API but I don't see an option to add workspace admin users.

**Reply by dhivah** [reply]
@frankie567 If I wanted to use multi tenant in Fief, currently it redirects to a subdomain but as I am running the service in GCP CloudRun. It is not possible to have multiple subdomains. I can add a load balancer to the CloudRun and forward the host header in `x-forwarded-host`. Would that work?

I am still unable to create additional workspace admin user from the UI and looks like the cli command `fief workspaces create-main-user` seems to be the only option but I cannot run CLI commands on a running service in CloudRun. what would be the best way to create additional users?

Can I configure SSO for admin dashboard access?

**Reply by frankie567** [reply]
Yes, currently, you can only add workspace administrators from the CLI. Alternatively, you can manually add a SQL row in the `workspace_users` table, with the right user and workspace ID.

If you want to use multiple workspaces in this context, yes, you'll need a reverse proxy that sets `x-forwarded-host`, it's supported. Don't forget to set the [`FORWARDED_ALLOW_IPS` environment variable to trust the proxy/load balancer](https://docs.fief.dev/self-hosting/deployment/ssl/#trust-reverse-proxy-headers).

**Reply by dhivah** [reply]
@frankie567 are there any plans in the roadmap to expose workspace operations as APIs? That would be very useful.

Can we set Google SSO for workspace administrator access?

**Reply by frankie567** [reply]
> @frankie567 are there any plans in the roadmap to expose workspace operations as APIs? That would be very useful.

Not currently, but why not!

> Can we set Google SSO for workspace administrator access?

Yes, if you configure an [OAuth Provider with Google](https://docs.fief.dev/configure/oauth-providers/) on your main workspace, it'll work as expected.

**Reply by dhivah** [reply]
@frankie567 I managed to create a new workspace but when I try going to the workspace it keeps redirecting to the fief main workspace URL. I am unable to find why. I created a separate database for the workspace and I can see the tables are created in that database. I also see an entry in the workspaces table on the main fief database. I am not sure why the redirect happens and I am unable to access the newly created workspace. Could you please tell me why?

**Reply by frankie567** [reply]
Can you try to clear all the cookies on your domain? It's possible that the admin session on your first workspace is maintained for some reasons.

**Reply by dhivah** [reply]
@frankie567 I tried configuring Google SSO as per the documentation but when I try to log in using Google SSO I get an error `Invalid login session.`. I don't see any further error messages in the log. I have cleared cookies and I am still getting the same error.

**Reply by dhivah** [reply]
@frankie567 Doesn't Fief support OAuth2 Client credentials grant type? This is an essential feature for me.

**Reply by frankie567** [reply]
> @frankie567 I tried configuring Google SSO as per the documentation but when I try to log in using Google SSO I get an error `Invalid login session.`. I don't see any further error messages in the log. I have cleared cookies and I am still getting the same error.

Yes, that's the same as bug #193!

> @frankie567 Doesn't Fief support OAuth2 Client credentials grant type? This is an essential feature for me.

Not right now, unfortunately; but that's planned!
